ios/browser/api/bookmarks/exporter/bookmark_html_writer.cc

```diff
@@ -0,0 +1,636 @@
+                            &mobile_folder_value) ||
+        mobile_folder_value->type() != base::Value::Type::DICTIONARY) {
+      NOTREACHED();  // Invalid type for root folder and/or other folder.
+    }
+
+    IncrementIndent();
+
+    if (!WriteNode(*static_cast<base::DictionaryValue*>(root_folder_value),
+                   BookmarkNode::BOOKMARK_BAR) ||
+        !WriteNode(*static_cast<base::DictionaryValue*>(other_folder_value),
+                   BookmarkNode::OTHER_NODE) ||
+        !WriteNode(*static_cast<base::DictionaryValue*>(mobile_folder_value),
+                   BookmarkNode::MOBILE)) {
+      NotifyOnFinish(BookmarksExportObserver::Result::kCouldNotWriteNodes);
+      return;
+    }
+
+    DecrementIndent();
+
+    Write(kFolderChildrenEnd);
+    Write(kNewline);
+    // File close is forced so that unit test could read it.
+    file_.reset();
+
+    NotifyOnFinish(BookmarksExportObserver::Result::kSuccess);
+  }
+
+ private:
+  friend class base::RefCountedThreadSafe<Writer>;
+
+  // Types of text being written out. The type dictates how the text is
+  // escaped.
+  enum TextType {
+    // The text is the value of an html attribute, eg foo in
+    // <a href="foo">.
+    ATTRIBUTE_VALUE,
+
+    // Actual content, eg foo in <h1>foo</h2>.
+    CONTENT
+  };
+
+  ~Writer() {}
+
+  // Opens the file, returning true on success.
+  bool OpenFile() {
+    int flags = base::File::FLAG_CREATE_ALWAYS | base::File::FLAG_WRITE;
+    file_.reset(new base::File(path_, flags));
+    if (!file_->IsValid()) {
+      PLOG(ERROR) << "Could not create " << path_;
+      return false;
+    }
+    return true;
+  }
+
+  // Increments the indent.
+  void IncrementIndent() {
+    indent_.resize(indent_.size() + kIndentSize, ' ');
+  }
+
+  // Decrements the indent.
+  void DecrementIndent() {
+    DCHECK(!indent_.empty());
+    indent_.resize(indent_.size() - kIndentSize, ' ');
+  }
+
+  // Called at the end of the export process.
+  void NotifyOnFinish(BookmarksExportObserver::Result result) {
+    if (observer_ != nullptr) {
+      observer_->OnExportFinished(result);
+    }
+  }
+
+  // Writes raw text out returning true on success. This does not escape
+  // the text in anyway.
+  bool Write(const std::string& text) {
+    if (!text.length())
+      return true;
+    size_t wrote = file_->WriteAtCurrentPos(text.c_str(), text.length());
+    bool result = (wrote == text.length());
+    if (!result) {
+      PLOG(ERROR) << "Could not write text to " << path_;
+      return false;
+    }
+    return true;
+  }
+
+  // Writes out the text string (as UTF8). The text is escaped based on
+  // type.
+  bool Write(const std::string& text, TextType type) {
+    DCHECK(base::IsStringUTF8(text));
+    std::string utf8_string;
+
+    switch (type) {
+      case ATTRIBUTE_VALUE:
+        // Convert " to &quot;
+        utf8_string = text;
+        base::ReplaceSubstringsAfterOffset(&utf8_string, 0, "\"", "&quot;");
```

Comment:
why not escape HTML in this case as well? for instance in `<a href=attribute>`, attribute could be `''></a><script src='evil.js'></script>`

Reply:
This is from Chromium itself: https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/bookmarks/bookmark_html_writer.cc;l=293;drc=929e790ad17a54008b3dac86cbe30751c968c2ce?q=html_writer&ss=chromium%2Fchromium%2Fsrc

Reply:
i guess their logic is that escaping quotes is sufficient because all the attribute values start with `"` instead of `'`